### PR TITLE
Refactor ConcurrentExecPool and related classes

### PR DESCRIPTION
**Pull request by bernhardmgruber**
This should be a pure refactoring with no functional changes. This is a preliminary refactoring to investigate the queuebug from https://github.com/alpaka-group/alpaka/pull/1850.

### CODE REVIEW

File: include/alpaka/core/ConcurrentExecPool.hpp [hunk 1]
```diff
@@ -33,41 +33,37 @@ namespace alpaka::core
...
             }
+
             //! Pushes the given value onto the back of the queue.
-            auto push(T&& t) -> void
+            void push(T&& t)
             {
-                std::lock_guard<std::mutex> lk(m_Mutex);
-
-                std::queue<T>::push(std::forward<T>(t));
+                std::lock_guard<std::mutex> lk(m_mutex);
```
Comment (fwyzard) [inline include/alpaka/core/ConcurrentExecPool.hpp:48]: Is the `<std::mutex>` part still needed with c++17 ?

File: include/alpaka/core/ConcurrentExecPool.hpp [hunk 2]
```diff
@@ -77,118 +73,74 @@ namespace alpaka::core
...
+            void runTask() noexcept
             {
                 try
                 {
                     run();
                 }
                 catch(...)
                 {
-// Workaround: Clang can not support this when natively compiling device code. See ConcurrentExecPool.hpp.
+// Workaround: Clang can not support this when natively compiling device code.
```
Comment (fwyzard) [inline include/alpaka/core/ConcurrentExecPool.hpp:89]: Just wondering: is this a problem only for CUDA, or also for HIP and SYCL device compilation ?

File: include/alpaka/core/ConcurrentExecPool.hpp (hunk 1 above)
Comment (bernhardmgruber) [reply include/alpaka/core/ConcurrentExecPool.hpp:48]: It shoudn't and I would prefer CTAD here, but clang gave me a warning (`-Wctad-maybe-unsupported`) so I provided the type explicitely.

File: include/alpaka/core/ConcurrentExecPool.hpp (hunk 2 above)
Comment (bernhardmgruber) [reply include/alpaka/core/ConcurrentExecPool.hpp:89]: I have no idea. I could even be that the workaround is no longer necessary. It was added in c7b27ed6afed86da26146d6ec3aacc3e350a1762 in May 2016.

File: include/alpaka/queue/QueueGenericThreadsNonBlocking.hpp [hunk 3]
```diff
@@ -73,11 +73,8 @@ namespace alpaka
                     m_dev.registerCleanup(
                         [pool = std::weak_ptr<ThreadPool>(m_workerThread)]() noexcept
                         {
-                            auto s = pool.lock();
-                            if(s)
-                            {
-                                s = s->takeDetachHandle();
-                            }
+                            if(auto s = pool.lock())
+                                static_cast<void>(s->takeDetachHandle()); // let returned shared_ptr destroy itself
```
Comment (j-stephan) [inline include/alpaka/queue/QueueGenericThreadsNonBlocking.hpp:77]:
```suggestion
                               std::ignore = s->takeDetachHandle(); // let returned shared_ptr destroy itself
```

File: include/alpaka/core/ConcurrentExecPool.hpp [hunk 4]
```diff
@@ -33,41 +33,37 @@ namespace alpaka::core
...
+
             [[nodiscard]] auto empty() const -> bool
             {
-                return std::queue<T>::empty();
+                return m_queue.empty();
             }
+
             //! Pushes the given value onto the back of the queue.
-            auto push(T&& t) -> void
+            void push(T&& t)
```
Comment (j-stephan) [inline include/alpaka/core/ConcurrentExecPool.hpp:46]: Is this the recommended practice now?

**Review by j-stephan — CHANGES_REQUESTED** [review]
Could you briefly describe why the refactoring is required at this point (or link to a relevant issue) in case we want to re-review this in the future?

**Comment by bernhardmgruber** [comment]
> Could you briefly describe why the refactoring is required at this point (or link to a relevant issue) in case we want to re-review this in the future?

Yes, this is a preliminary refactoring to investigate the queuebug from #1850.

File: include/alpaka/queue/QueueGenericThreadsNonBlocking.hpp (hunk 3 above)
Comment (bernhardmgruber) [reply include/alpaka/queue/QueueGenericThreadsNonBlocking.hpp:77]: I know you can assign anything to `std::ignore`, but I find this even more arcane than the established cast to `void`. Would you prefer assigning to `std::ignore`? I don't care about it.

File: include/alpaka/core/ConcurrentExecPool.hpp (hunk 4 above)
Comment (bernhardmgruber) [reply include/alpaka/core/ConcurrentExecPool.hpp:46]: I only use a trailing return type for non-void return types and so far nobody complained.

File: include/alpaka/core/ConcurrentExecPool.hpp (hunk 2 above)
Comment (j-stephan) [reply include/alpaka/core/ConcurrentExecPool.hpp:89]: Please open an issue for the work-around. We should try to remove it.

**Review by j-stephan — APPROVED** [review]

**Pull request merged**